Sources/NIOTestUtils/NIOHTTP1TestServer.swift

```diff
@@ -0,0 +1,318 @@
+///     let allocator = ByteBufferAllocator()
+///     let testServer = NIOHTTP1TestServer(group: group)
+///     defer {
+///         XCTAssertNoThrow(try testServer.stop())
+///         XCTAssertNoThrow(try group.syncShutdownGracefully())
+///     }
+///
+///     // Use your library to send a request to the server.
+///     let requestBody = "ping"
+///     var requestComplete: EventLoopFuture<String>!
+///     XCTAssertNoThrow(requestComplete = try sendRequestTo(
+///         URL(string: "http://127.0.0.1:\(testServer.serverPort)/some-route")!,
+///         body: requestBody))
+///
+///     // Assert the server received the expected request.
+///     // Use custom methods if you only want some specific assertions on part
+///     // of the request.
+///     XCTAssertNoThrow(XCTAssertEqual(.head(.init(version: .init(major: 1, minor: 1),
+///                                                 method: .GET,
+///                                                 uri: "/some-route",
+///                                                 headers: .init([
+///                                                     ("Content-Type", "text/plain; charset=utf-8"),
+///                                                     ("Content-Length", "4")]))),
+///                                     try testServer.readInbound()))
+///     var requestBuffer = allocator.buffer(capacity: 128)
+///     requestBuffer.writeString(requestBody)
+///     XCTAssertNoThrow(XCTAssertEqual(.body(requestBuffer),
+///                                     try testServer.readInbound()))
+///     XCTAssertNoThrow(XCTAssertEqual(.end(nil),
+///                                     try testServer.readInbound()))
+///
+///     // Make the server send a response to the client.
+///     let responseBody = "pong"
+///     var responseBuffer = allocator.buffer(capacity: 128)
+///     responseBuffer.writeString(responseBody)
+///     XCTAssertNoThrow(try testServer.writeOutbound(.head(.init(version: .init(major: 1, minor: 1), status: .ok))))
+///     XCTAssertNoThrow(try testServer.writeOutbound(.body(.byteBuffer(responseBuffer))))
+///     XCTAssertNoThrow(try testServer.writeOutbound(.end(nil)))
+///
+///     // Assert that the client received the response from the server.
+///     XCTAssertNoThrow(XCTAssertEqual(responseBody, try requestComplete.wait()))
+public final class NIOHTTP1TestServer {
+    private let eventLoop: EventLoop
+    // all protected by eventLoop
+    private let inboundBuffer: BlockingQueue<HTTPServerRequestPart> = .init()
+    private var currentClientChannel: Channel? = nil
+    private var serverChannel: Channel! = nil
+
+    enum State {
+        case channelsAvailable(CircularBuffer<Channel>)
+        case waitingForChannel(EventLoopPromise<Void>)
+        case idle
+        case stopped
+    }
+    private var state: State = .idle
+
+    func handleChannels() {
+        self.eventLoop.assertInEventLoop()
+
+        let channel: Channel!
```

Comment:
This doesn’t need to be an IUO, there must be a `Channel` if we get past this section.